plugins/tc/tbf/tbf_suite_test.go

```diff
@@ -0,0 +1,193 @@
+package main
+
+// Copyright 2015 CNI authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"bufio"
+	"bytes"
+	"fmt"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/vishvananda/netlink"
+	"net"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestTBF(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "tbf suite")
+}
+
+func makeTcpClientInNS(hostNs ns.NetNS, addr string, numBytes int) {
+	Expect(hostNs.Do(func(netNS ns.NetNS) error {
+		conn, err := net.DialTimeout("tcp", addr, 2*time.Second)
+		if err != nil {
+			return err
+		}
+		defer conn.Close()
+
+		data := append(bytes.Repeat([]byte{'a'}, numBytes), '\n')
+
+		conn.SetWriteDeadline(time.Now().Add(time.Minute))
+		conn.SetReadDeadline(time.Now().Add(time.Minute))
+
+		_, err = conn.Write(data)
+		if err != nil {
+			return err
+		}
+
+		reply := make([]byte, 1)
+		_, err = conn.Read(reply)
+
+		return err
+	})).To(Succeed())
+}
+
+func startTcpServerInNS(containerNS ns.NetNS, laddr string) {
```

Comment:
At a glance, this sounds a lot like the [`echosvr` we use in the `portmap` integration tests](https://github.com/containernetworking/plugins/blob/92c634042c3888d942199c396c1d6f42d82785f3/plugins/meta/portmap/portmap_suite_test.go#L66-L78).

Perhaps there's an opportunity for code-reuse?

Reply:
Ah, thanks for pointing this out. I like the idea of code-reuse. 

I am finding using echosvr a bit tricky to use for this particular test. Mainly because the echosvr [tries to read a fixed amount of bytes](https://github.com/containernetworking/plugins/blob/92c634042c3888d942199c396c1d6f42d82785f3/pkg/testutils/echosvr/main.go#L34) and then continues to write back what it read and close the connection. However in the measure case I want to ensure that the entire message is sent and read by the server.  I did this previously by sending a message with a trailing new-line character as a signal to the server that the entire message had been sent, and that it could then write back a response so the client could know that the server received the entire message.

I have tried to get this to work by tweaking the num of packets sent to the 'echo server' and tbf's burst/limit sent to fit into the buffer size of 512. However this results in the time taken with/without tbf to be negligible (~100 millisecond difference) 

I have tried to increase the `echosvr` buffer size  to allow sending/reading a larger payload, but `conn.Read` seems like it reads as much as the `read syscall` is set to (which when I run is consistently 61232 bytes). If i send < 61232 to the `echosvr` it results in a `broken pipe error`. Haven't looked too much into it but I suspect that since ingress and egress traffic are both shaped, that when the `echo svr`  tries to write back a large payload it gets shaped, and it closes the connection prematurely.

Your thoughts are much appreciated.

Reply:
I'd be happy with replacing the echoserver with this code. It really should be in another process, if only because Go namespace crappiness has caused us a lot of grief and flaky tests.

Reply:
ok thanks for the reply. will do